build_tools/docker/bazel/Dockerfile

```diff
@@ -27,40 +27,43 @@
 FROM ubuntu:18.04
 WORKDIR /usr/src/iree
 
-RUN apt-get update
-
 # Set environment variables.
 ENV CXX clang++
 ENV CC clang
 ENV PYTHON_BIN /usr/bin/python3
 ENV IREE_LLVMAOT_LINKER_PATH /usr/bin/ld
 
-# Install git for updating IREE's submodules.
-RUN apt-get install -y git
+RUN apt-get update \
+  && apt-get install -y \
+    # git for updating IREE's submodules.
+    git \
+    # utilities for later installations
+    unzip \
+    zip \
+    wget \
+    # core IREE dependencies.
+    clang \
+    libsdl2-dev
+
+# Disable apt-key parse waring. If someone knows how to do whatever the "proper"
+# thing is then feel free. The warning complains about parsing apt-key output,
+# which we're not even doing.
+ARG APT_KEY_DONT_WARN_ON_DANGEROUS_USAGE=1
 
 # Install Bazel.
 # https://docs.bazel.build/versions/master/install-ubuntu.html
-ARG BAZEL_VERSION=2.1.0
-RUN apt-get install -y unzip zip wget \
-  && wget "https://github.com/bazelbuild/bazel/releases/download/${BAZEL_VERSION?}/bazel-${BAZEL_VERSION?}-installer-linux-x86_64.sh" \
-  && chmod +x "bazel-${BAZEL_VERSION?}-installer-linux-x86_64.sh" \
-  && "./bazel-${BAZEL_VERSION?}-installer-linux-x86_64.sh" --user \
-  && rm "bazel-${BAZEL_VERSION?}-installer-linux-x86_64.sh"
-# Install a newer version of Bazel. We don't need the full installation now.
-# Just need to provide a different version for the version-identification
-# wrapper script to find in /root/.bazel/bin
+ARG BAZEL_VERSION=3.3.1
+# Change to a new version if upgrading Bazel.
 ARG NEW_BAZEL_VERSION=3.3.1
-RUN cd "/root/.bazel/bin" \
-  && wget "https://releases.bazel.build/${NEW_BAZEL_VERSION?}/release/bazel-${NEW_BAZEL_VERSION?}-linux-x86_64" \
-  && chmod +x "bazel-${NEW_BAZEL_VERSION?}-linux-x86_64"
-# ENV does not allow ${variable?} syntax.
-ENV PATH "/root/bin:${PATH}"
-
-# Install core IREE dependencies.
-RUN apt-get install -y clang libsdl2-dev
+RUN wget -qO - https://bazel.build/bazel-release.pub.gpg | apt-key add - \
+  && echo "deb [arch=amd64] https://storage.googleapis.com/bazel-apt stable jdk1.8" \
+    | tee /etc/apt/sources.list.d/bazel.list \
+  && apt-get update \
+  && apt-get install "bazel=${BAZEL_VERSION?}" "bazel-${NEW_BAZEL_VERSION?}"
```

Comment:
It's actually installing the same version twice :-)

When upgrading Bazel, we'll always need to temporarily have both the old and new versions to allow for builds to succeed before and after the `.bazelversion` bump. Bazel does some magic where the `bazel` binary is actually a shell script that finds the right Bazel binary based on `.bazelversion`. The exact pattern here is fiddly enough that I thought it wasn't bad to just leave it in, since the actual cost of installing the same thing twice is very low, I think. Then the next time we need to upgrade, someone doesn't have to figure this out again. Could also be a comment, but this seemed easier at low cost. Other suggestions?

Reply:
Ah, I see. This is very subtle and would worth a comment. :)